### ISSUE TITLE
Slots don't converge for object-wise decomposition

### ISSUE THREAD

**Issue by franciscocms** [davzha/MESH#2]
Hey! First of all, thank you for having the code available! I am using your SA-MESH variant of slot attention for a slightly different problem, where the training signal that is backpropagated is not the reconstruction error (neither `reg_loss`), but the summed log probability of the true latent variables in a posterior distribution with proposed parameters. In this case, even though slots are sparse, I cannot make them converge for object-centric representations. Do you have any insights about the generality of SA-MESH when it comes to different optimization signals? Thank you in advance

**Comment by davzha** [comment]
If I understand correctly your aim is to directly supervise the slots? That is what we do in the first experiment of the paper. Are you using a permutation invariant loss (e.g., Hungarian algorithm)?

**Comment by franciscocms** [comment]
Thanks for your reply! Exactly, the main difference is that I am not allowing some permutation in my inferred object ordering. This is due to the fact that I am using SA-MESH to learn proposals for the posterior distribution of latent variables in a probabilistic program, hence I need to specify for which object (i.e. property of object) I am proposing a distribution for... (I hope I gave you a clear view!) - even though the object indexes in the probabilistic program respect the euclidean distance to the origin of the image (so that the inference network can learn to assign first slots to objects that are closer to the origin), what is happening is that the system tends to converge for a local minimum where it doesn't see benefits on exclusively assigning slots to objects, and slots representations keep encoding information over all objects (in different regions, because some sparsity is still enforced due to the mesh algorithm)...

**Comment by davzha** [comment]
This sounds like the issue might be with the permutation symmetry. Have you tried to initiate the slots with their respective object properties (or an affine transformation thereof)?

**Comment by franciscocms** [comment]
@davzha yeah, the problem was definitely with the permutation symmetry, just tried that out and works like a charm!
I'm closing the issue then, thank you!
